x-pack/solutions/security/packages/index-adapter/src/reindex_index.ts

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient, Logger } from '@kbn/core/server';
+import { retryTransientEsErrors } from './retry_transient_es_errors';
+
+export interface ReindexIndexDocumentsParams {
+  esClient: ElasticsearchClient;
+  logger: Logger;
+  indexName: string;
+  batchSize?: number;
+  timeout?: string;
+}
+
+export interface ReindexTaskStatus {
+  taskId: string;
+  completed: boolean;
+  error?: any;
+  total?: number;
+  created?: number;
+  updated?: number;
+  batches?: number;
+}
+
+/**
+ * Creates a new index with updated mappings and reindexes documents from the old index
+ * This is useful when updating inference_id properties of semantic_text fields
+ */
+export async function reindexIndexDocuments({
+  esClient,
+  logger,
+  indexName,
+  batchSize = 1000,
+  timeout = '10m',
+}: ReindexIndexDocumentsParams): Promise<void> {
+  logger.info(`Starting reindex of documents in index: ${indexName}`);
+
+  try {
+    // Check if index exists
+    const indexExists = await retryTransientEsErrors(
+      () => esClient.indices.exists({ index: indexName }),
+      { logger }
+    );
+
+    if (!indexExists) {
+      throw new Error(`Index ${indexName} does not exist`);
+    }
+
+    // Get current index settings and mappings
+    const indexInfo = await retryTransientEsErrors(
+      () => esClient.indices.get({ index: indexName }),
+      { logger }
+    );
+
+    const currentIndex = indexInfo[indexName];
+    if (!currentIndex) {
+      throw new Error(`Could not retrieve information for index ${indexName}`);
+    }
+
+    // Create temporary index name
+    const tempIndexName = `${indexName}-reindex-${Date.now()}`;
+
+    // Create new index with updated mappings from template
+    await createIndexWithUpdatedMappings({
+      esClient,
+      logger,
+      sourceIndexName: indexName,
+      tempIndexName,
+    });
+
+    // Reindex documents to the new index
+    await performReindex({
+      esClient,
+      logger,
+      sourceIndex: indexName,
+      targetIndex: tempIndexName,
+      batchSize,
+      timeout,
+    });
```

Comment:
This can take a long time if there are a lot of documents. Is there anything preventing additional reIndex operations from being kicked off and causing issues?